libs/types-common/src/Data/Json/Util.hs

```diff
@@ -42,18 +48,37 @@ infixr 5 #
 
 -- | A newtype wrapper for 'UTCTime' that formats timestamps in JSON with
 -- millisecond precision instead of the default picosecond precision.
+-- Construct values using 'toUTCTimeMillis'; deconstruct with 'fromUTCTimeMillis'.
+-- Unlike with 'UTCTime', 'Show' renders ISO string.
 newtype UTCTimeMillis = UTCTimeMillis { fromUTCTimeMillis :: UTCTime }
-  deriving (Eq)
 
-instance Show UTCTimeMillis where
-    showsPrec d = showParen (d > 10) . showString . showUTCTimeMillis
+{-# INLINE toUTCTimeMillis #-}
+toUTCTimeMillis :: HasCallStack => UTCTime -> UTCTimeMillis
+toUTCTimeMillis = UTCTimeMillis . (TL.seconds %~ MkFixed . roundit . coerce @Pico @Integer)
+  where
+    roundit = (* onebillion) . (`div` onebillion)
+    onebillion = 10 ^ (9 :: Integer)
 
 {-# INLINE showUTCTimeMillis #-}
 showUTCTimeMillis :: UTCTimeMillis -> String
-showUTCTimeMillis (UTCTimeMillis t) = formatTime defaultTimeLocale format t
-  where
-    format = "%FT%T." ++ formatMillis t ++ "Z"
-    formatMillis = take 3 . formatTime defaultTimeLocale "%q"
+showUTCTimeMillis (UTCTimeMillis t) = formatTime defaultTimeLocale formatUTCTimeMillis t
+
+readUTCTimeMillis :: String -> Maybe UTCTimeMillis
+readUTCTimeMillis = fmap toUTCTimeMillis . parseTimeM True defaultTimeLocale formatUTCTimeMillis
+
+formatUTCTimeMillis :: String
+formatUTCTimeMillis = "%FT%T%QZ"
+
+{-# INLINE eqUTCTimeMillis #-}
+-- (this implementation is not very fast)
```

Comment:
Wouldn't it be better to do something like this?

``` haskell
eqUTCTimeMillis = (==) `on` (roundToMillis . fromUTCTimeMillis)
```

Where `roundToMillis :: UTCTime -> UTCTime` would be an internal helper function.

Reply:
https://github.com/wireapp/wire-server/pull/366/commits/8f68b934da298907c3bdb3a5a2c6de2f01aeb688